tools/automate_gcp_os_version.py

```diff
@@ -0,0 +1,118 @@
+"""
+checks for the latest os and updates the os versions in vars.tf file in dc-only, 
+multi-region and single connector deployments.
+Use this command to run this script in cloud shell: python3 tools/update_gcp_os.py
+"""
+import argparse
+import datetime
+import os
+import re
+import subprocess
+from subprocess import CalledProcessError
+
+COMPUTE_SERVICES_ENABLE = "gcloud services enable compute.googleapis.com"
+COMMANDS = ["gcloud compute images list"]
+PATHS = [["single-connector", "vars.tf"],
+         ["multi-region", "vars.tf"],
+         ["dc-only", "vars.tf"]]
+
+# commands = ["gcloud compute images list | grep windows-server-2016",
+#             'gcloud compute images list | grep centos-7',
+#             'gcloud compute images list | grep ubuntu-1804-bionic']
+
+
+# Added env variable for future purpose
+def get_os(env='cloud'):
+    """
+    get the latest version from GCP
+    """
+    try:
+        recent_os = []
+        if env == "cloud":
+            # breakpoint()
+            for command in COMMANDS:
+                capture_output = subprocess.run(command.split(' '), check=True,
+                                                capture_output=True).stdout.decode().split('\n')
+                recent_os.extend([key[0] for key in map(lambda x: x.split(' '), capture_output)])
+        return recent_os
+    except (AttributeError, TypeError, CalledProcessError) as ex:
+        print("Exception occurred: ", ex)
+
+
+def compare_versions(recent_os, line):
+    """
+    Checks for version numbers and returns true if latest version available
+    """
+    try:
+        # version_date variable is the current os version in the vars.tf
+        version = re.compile(r'[0-9]{8}').search(line.split("/")[-1])
+        if version:
+            version = version.group()
+            # print("version", version)
+            version_date = datetime.datetime.strptime(version, "%Y%m%d")
+            for _os in recent_os:
+                latest_gcp_os = re.compile(r'[0-9]{8}').search(_os)
+                # print(latest_gcp_os)
+                if latest_gcp_os:
```

Comment:
with the above changes, hope you don't need these OS-specific if statements anymore and the code will work with all OSes, assuming they have the same naming pattern. (everything other than cos-cloud)

Reply:
Only one if condition removed, but os specific conditions needed, we don't know which line contains which OS in vars.tf, so we are comparing the each line in the vars.tf with the existing OS'es with available latest OS versions in GCP.

Reply:
We're identifying lines that correspond to image by first looking for "global/images".  Once we have a match, we know that whatever comes after "global/images/" and before "-vYYYYMMDD" is the OS we are looking for.  So we could re-use that info to look up entries in the recent_os list. This would work for all GCP images except "cos-cloud" one which we don't use.  The code then doesn't need to have specific cases to handle different OSes.